presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,552 @@
+         * @param valueOnlyRawHash
+         * @param nodePointerToUse
+         * @param isEmpty
+         */
+        private BucketDataNode(int bucketId, ValueNode valueNode, long valueRawHash, long valueOnlyRawHash, int nodePointerToUse, boolean isEmpty)
+        {
+            this.bucketId = bucketId;
+            this.valueNode = valueNode;
+            this.valueRawHash = valueRawHash;
+            this.valueOnlyRawHash = valueOnlyRawHash;
+            this.nodePointerToUse = nodePointerToUse;
+            this.isEmpty = isEmpty;
+        }
+
+        private boolean isEmpty()
+        {
+            return isEmpty;
+        }
+
+        /**
+         * true iff needs to update nextNodePoint
+         *
+         * @param groupId
+         * @param block
+         * @param position
+         * @param count
+         * @return
+         */
+        private boolean processEntry(long groupId, Block block, int position, long count)
+        {
+            if (isEmpty()) {
+                addNewGroup(groupId, block, position, count);
+                return true;
+            }
+            else {
+                valueNode.add(count);
+                return false;
+            }
+        }
+
+        private void addNewGroup(long groupId, Block block, int position, long count)
+        {
+            checkState(isEmpty(), "bucket %s not empty, points to %s", bucketId, buckets.get(bucketId));
+
+            // we've already computed the value hash for only the value only; ValueStore will save it for future use
+            int nextValuePosition = valueStore.getPosition(type, block, position, valueOnlyRawHash);
+            valuePositions.set(nodePointerToUse, nextValuePosition);
+            // save hashes for future rehashing
+            valueRawHashList.set(nodePointerToUse, valueRawHash);
+            // set bucket
+            buckets.set(bucketId, nodePointerToUse);
+            // save data
+            counts.set(nodePointerToUse, count);
+            // used for doing value comparisons on hash collisions
+            groupIds.set(nodePointerToUse, groupId);
+            // we only ever store ints as values; we need long as an index
+            int currentHead = (int) headPointers.get(groupId);
+
+            headPointers.set(groupId, nodePointerToUse);
+            nextPointers.set(nodePointerToUse, currentHead);
+        }
+    }
+
+    private class BucketNodeFactory
+    {
+        /**
+         * invariant: "points" to a virtual node of [key, count] for the histogram and includes any indirection calcs. Makes not guarantees if the node is empty or not
+         * (use isEmpty())
+         *
+         * @param currentBuckets
+         * @param i
+         * @param groupId
+         * @param block
+         * @param position
+         */
+        private BucketDataNode createBucketDataNode(long groupId, Block block, int position)
+        {
+            long valueOnlyRawHash = TypeUtils.hashPosition(type, block, position);
```

Comment:
Maybe call this valueHash and valueAndGroupHash for the one below.

Reply:
done